packages/blob/src/index.ts

```diff
@@ -215,13 +231,152 @@ export async function list(
   };
 }
 
-function getApiUrl(): string {
-  return process.env.VERCEL_BLOB_API_URL || 'https://blob.vercel-storage.com';
+export interface GenerateClientTokenOptions extends BlobCommandOptions {
+  pathname: string;
+  onUploadCompleted?: {
+    callbackUrl: string;
+    metadata?: string;
+  };
+  maximumSizeInBytes?: number;
+  allowedContentTypes?: string[];
+}
+
+export async function generateClientTokenFromReadWriteToken({
+  token,
+  ...args
+}: GenerateClientTokenOptions): Promise<string> {
+  if (typeof window !== 'undefined') {
+    throw new Error(
+      '"generateClientTokenFromReadWriteToken" must be called from a server environment',
+    );
+  }
+  const timestamp = new Date();
+  timestamp.setSeconds(timestamp.getSeconds() + 30);
+  const blobToken = getToken({ token });
+
+  const [, , , storeId = null] = blobToken.split('_');
+
+  if (!storeId) {
+    throw new Error(
+      token ? 'Invalid "token" parameter' : 'Invalid BLOB_READ_WRITE_TOKEN',
+    );
+  }
+
+  const payload = Buffer.from(
+    JSON.stringify({
+      ...args,
+      validUntil: timestamp.getTime(),
+    }),
+  ).toString('base64');
+
+  const securedKey = await signPayload(payload, blobToken);
+  if (!securedKey) {
+    throw new Error('Unable to sign client token');
+  }
+  return `vercel_blob_client_${storeId}_${Buffer.from(
+    `${securedKey}.${payload}`,
+  ).toString('base64')}`;
+}
+
+async function importKey(token?: string): Promise<CryptoKey> {
+  return globalThis.crypto.subtle.importKey(
+    'raw',
+    new TextEncoder().encode(getToken({ token })),
+    { name: 'HMAC', hash: 'SHA-256' },
+    false,
+    ['sign', 'verify'],
+  );
 }
 
-function getToken(putOptions?: BlobCommandOptions): string {
-  if (putOptions?.token) {
-    return putOptions.token;
+async function signPayload(
+  payload: string,
+  token: string,
+): Promise<string | undefined> {
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  if (!globalThis.crypto) {
+    return crypto.createHmac('sha256', token).update(payload).digest('hex');
+  }
+
+  const signature = await globalThis.crypto.subtle.sign(
+    'HMAC',
+    await importKey(token),
+    new TextEncoder().encode(payload),
+  );
+  return Buffer.from(new Uint8Array(signature)).toString('hex');
+}
+
+export async function verifyCallbackSignature({
+  token,
+  signature,
+  body,
+}: {
+  token?: string;
+  signature: string;
+  body: string;
+}): Promise<boolean> {
+  const secret = getToken({ token });
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
```

Comment:
If we can add a comment here too that would be nice (not obvious why we're doing this)